Validate requests

khieta opened the pull request:
[rendered](https://github.com/cedar-policy/rfcs/blob/rfc/khieta/validate-queries/text/0011-validate-requests.md)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

mwhicks1 commented on text/0011-validate-requests.md:
```diff
@@ -0,0 +1,72 @@
+## Basic example
+
+Say the schema specifies that an action `readFile` should only be used with principals of type `User` and resources of type `File`.
+This RFC proposes to change the signature of the `Request` constructor to add an optional schema argument that can be used to check  argument types when constructing a `Request`.
```
Do we have to change the signature of the constructor? Can we add a new constructor instead, with the schema parameter added (whereas the old one stays as is) ?

mwhicks1 commented on text/0011-validate-requests.md:
````diff
@@ -0,0 +1,72 @@
+```
+
+Assuming that they pass in inputs where the principal is of type `User` and the resource is of type `File`, this will work as expected.
+However, it could also return "allow" for other principal types like `Folder` or resource types like `User`.
````
This feels contrived. Why would I have a `Folder` principal type when using the `readFile` action? Moreover, why does `User` as a resource type matter in this case? Presumably no `User` entities exist in `Folder::"public"`. Is there a way to tweak your "for example" to make the chances of a bug occurring seem more plausible?

khieta replied:
I guess a more natural mistake in this setup would for the resource to be a `Folder`. In this case, the result of the policy could be "allow" (assuming that folders can be nested). Is this a more convincing use case?

I'll do some thinking to come up with a better example. Open to suggestions in the meantime.

mwhicks1 replied:
I like the `Folder` use-case. Then you'd have to change your example, though, to use `is` for the resource, rather than the principal, right?

mwhicks1 commented on text/0011-validate-requests.md:
```diff
@@ -0,0 +1,72 @@
+
+## Detailed design
+
+Validation and schema-based parsing are both "opt-in" in the sense that they are not required, so we'll the same for query validation.
```
Typo ("we'll the same for")

mwhicks1 commented on text/0011-validate-requests.md:
```diff
@@ -0,0 +1,72 @@
+
+## Drawbacks
+
+- This is a breaking change that will affect any users that are manually constructing `Request`s (which is likely all users).
```
Does it have to be? For example, we have a `new_with_unknowns` constructor (https://github.com/cedar-policy/cedar/blob/f233431efb82a1323e35cba2023b34613b8c99e5/cedar-policy-core/src/ast/request.rs#L91); could we have a `new_with_validation` constructor? 

For that matter -- should this RFC apply to partial evaluation-ready `Request`s too?

khieta replied:
To answer your second question: yes, I think we'll want validation supported for partial `Request`s too, once those are exposed.

mwhicks1 replied:
> The downside is that this will break existing users; the benefit is that this will force them to be aware of the new feature and start validating their requests.

Good characterization of the tradeoff. IMO, I'd rather go with two APIs because I don't want to break existing users, and because I expect that the non-validation path is the common one. People will largely use the validation path for debugging.

How would you feel about having two `new` methods, one with 4 parameters and one with 5? (I'm pretty sure Rust supports overloading.) The five-parameter one would call validation, and then invoke the four-parameter one.

shaobo-he-aws replied:
IIRC, Rust doesn't support function overloading.

cdisselkoen replied:
Rust does not support function overloading.  You can use traits to give the same function different definitions on different types, but the signature has to match with the exception of the type `T` that's implementing the trait.

john-h-kastner-aws replied:
I somewhat in favor of making the breaking change.

cdisselkoen replied:
Same.  It seems like a worthwhile change to force people to adapt to when updating to Cedar 3.x (whenever 3.x happens).

cdisselkoen replied:
I'm imagining that the actual performance overhead of request validation would be negligible, noting that the issue of validating the `context` is orthogonal (users can already do so or not as they construct the `Context` object which is passed to `Request::new()`), so the only performance overhead would be checking the `principal`, `action`, and `resource` are of compatible types. (Right?)

mwhicks1 replied:
To clarify: Users will validate the context if they do schema-based parsing to construct. They aren't required to do that.

cdisselkoen replied:
Correct.  It's orthogonal to whether they validate the request (which would only happen if the supply the schema to `Request::new()`).  They could do either one alone, or both, or neither.

shaobo-he-aws commented on text/0011-validate-requests.md:
```diff
@@ -0,0 +1,72 @@
+
+## Motivation
+
+(Example migrated from #5)
```
It seems #5 doesn't have a hyperlink when rendered.

khieta commented:
I've revised the RFC and am looking for new reviews. Summary of changes:
1. Revised the User/File/Folder example per Mike's suggestion.
2. Added an additional example to the motivation section to show why validating requests is important in conjunction with standard validation.
3. Added an "unresolved question" about changing the `Request::new` API vs. adding a new `Request::new_with_validation` API.

Please weigh in on point (3) if you haven't already.

khieta commented:
Revised! Notable changes:
* Per in-person discussion yesterday, instead of modifying the `Request::new` constructor, we will add a new `Request::new_with_validation` constructor.
* Added more concrete details about implementation (copied below for reference).

> The new API will have the following signature:
> 
> ```rust
> pub fn new_with_validation(
>         principal: Option<EntityUid>,
>         action: Option<EntityUid>,
>         resource: Option<EntityUid>,
>         context: Context,
>         schema: &Schema
>     ) -> Result<Request> {
>         ...
>     }
> ```
> 
> (This is the same as the signature for `Request::new`, aside from the schema argument and return type.)
> 
> If the action is `None` (indicating that it is "unspecified"), then the function performs no additional checks. If the action is `Some`, then the function checks that the specified action is present in the schema, and that the principal and resource are consistent with the action's `appliesTo` lists in the schema.

mwhicks1 commented on text/0011-validate-requests.md:
```diff
@@ -20,16 +20,16 @@ This RFC proposes to add support for checking that a request conforms to a schem
+Request::new_with_validation(Some(User::"alice"), Some(Action::"readFile"), Some(File::"secret_file.txt"), context, Some(schema)); // returns a Request
 
-Request::new(Some(User::"alice"), Some(Action::"readFile"), Some(Folder::"some_folder"), context, Some(schema)); // error, invalid resource
+Request::new_with_validation(Some(User::"alice"), Some(Action::"readFile"), Some(Folder::"some_folder"), context, Some(schema)); // returns an error (invalid resource)
```
Drop the `Some(...)` part here, since the schema is required for this new constructor.

khieta commented:
Based on discussion, we generally support the addition of the `Request::new_with_validation` constructor. However, I've decided that this change does not merit the bar for an RFC, so rather than moving into FCP, I will be closing the issue. The core content of the RFC will be moved to an issue in the `cedar` repository.

khieta commented:
Moved to `cedar` [issue #191](https://github.com/cedar-policy/cedar/issues/191).
